File: infra/src/main/java/com/beat/infra/persistence/booking/repository/BookingRepositoryImpl.java
package com.beat.infra.persistence.booking.repository;

import java.util.List;
import java.util.Optional;

import org.springframework.stereotype.Repository;

import com.beat.domain.booking.dao.BookingRepository;
import com.beat.domain.booking.domain.Booking;
import com.beat.domain.booking.domain.BookingStatus;
import com.beat.infra.persistence.booking.entity.BookingJpaEntity;
import com.beat.infra.persistence.booking.mapper.BookingPersistenceMapper;

@Repository
public class BookingRepositoryImpl implements BookingRepository {

	private final BookingJpaRepository bookingJpaRepository;
	private final BookingPersistenceMapper bookingPersistenceMapper;

	public BookingRepositoryImpl(BookingJpaRepository bookingJpaRepository,
		BookingPersistenceMapper bookingPersistenceMapper) {
		this.bookingJpaRepository = bookingJpaRepository;
		this.bookingPersistenceMapper = bookingPersistenceMapper;
	}

	@Override
	public Booking save(Booking booking) {
		BookingJpaEntity savedEntity = bookingJpaRepository.save(bookingPersistenceMapper.toEntity(booking));
		return bookingPersistenceMapper.toDomain(savedEntity);
	}

	@Override
	public Optional<Booking> findById(Long id) {
		return bookingJpaRepository.findById(id).map(bookingPersistenceMapper::toDomain);
	}

	@Override
	public List<Booking> findAll() {
		return bookingJpaRepository.findAll().stream()
			.map(bookingPersistenceMapper::toDomain)
			.toList();
	}

	@Override
	public Optional<List<Booking>> findByBookerNameAndBookerPhoneNumberAndPasswordAndBirthDate(
		String bookerName,
		String bookerPhoneNumber,
		String password,
		String birthDate
	) {
		return bookingJpaRepository.findByBookerNameAndBookerPhoneNumberAndPasswordAndBirthDate(
			bookerName,
			bookerPhoneNumber,
			password,
			birthDate
		).map(entities -> entities.stream()
			.map(bookingPersistenceMapper::toDomain)
			.toList());
	}

	@Override
	public Optional<Booking> findFirstByBookerNameAndBookerPhoneNumberAndBirthDateAndPassword(
		String bookerName,
		String bookerPhoneNumber,
		String birthDate,
		String password
	) {
		return bookingJpaRepository.findFirstByBookerNameAndBookerPhoneNumberAndBirthDateAndPassword(
			bookerName,
			bookerPhoneNumber,
			birthDate,
			password
		).map(bookingPersistenceMapper::toDomain);
	}

	@Override
	public List<Booking> findByUserId(Long userId) {
		return bookingJpaRepository.findByUserId(userId).stream()
			.map(bookingPersistenceMapper::toDomain)
			.toList();
	}

	@Override
	public boolean existsActiveBookingByScheduleIds(List<Long> scheduleIds, List<BookingStatus> excludedStatuses) {
		if (scheduleIds == null || scheduleIds.isEmpty()) {
			return false;
		}
		return bookingJpaRepository.existsActiveBookingByScheduleIds(scheduleIds, excludedStatuses);
	}

	@Override
	public int deleteInactiveBookingsByScheduleIds(List<Long> scheduleIds, List<BookingStatus> inactiveStatuses) {
		if (scheduleIds == null || scheduleIds.isEmpty()) {
			return 0;
		}
		return bookingJpaRepository.deleteInactiveBookingsByScheduleIds(scheduleIds, inactiveStatuses);
	}
}

File: core/infra/src/main/kotlin/com/beat/infra/persistence/booking/repository/BookingRepositoryImpl.kt
package com.beat.infra.persistence.booking.repository

import com.beat.domain.booking.model.Booking
import com.beat.domain.booking.model.BookingStatus
import com.beat.domain.booking.repository.BookingRepository
import com.beat.infra.persistence.booking.mapper.BookingPersistenceMapper
import org.springframework.stereotype.Repository
import java.time.LocalDateTime
import java.util.Optional

@Repository
class BookingRepositoryImpl(
    private val bookingJpaRepository: BookingJpaRepository,
    private val bookingPersistenceMapper: BookingPersistenceMapper,
) : BookingRepository {
    override fun save(booking: Booking): Booking =
        bookingPersistenceMapper.toDomain(
            bookingJpaRepository.save(bookingPersistenceMapper.toEntity(booking)),
        )

    override fun findById(id: Long?): Optional<Booking> =
        bookingJpaRepository.findById(requireNotNull(id) { "The given id must not be null" })
            .map(bookingPersistenceMapper::toDomain)

    override fun findScheduleIdsByIds(ids: Collection<Long>): List<Long> {
        if (ids.isEmpty()) {
            return emptyList()
        }
        return bookingJpaRepository.findScheduleIdsByIds(ids)
    }

    override fun lockById(id: Long?): Optional<Booking> =
        bookingJpaRepository.lockById(id).map(bookingPersistenceMapper::toDomain)

    override fun findAll(): List<Booking> =
        bookingJpaRepository.findAll().map(bookingPersistenceMapper::toDomain)

    override fun deleteAll(bookings: Iterable<Booking>) {
        bookingJpaRepository.deleteAll(bookings.map(bookingPersistenceMapper::toEntity))
    }

    override fun findByBookingStatusAndCancellationDateBefore(
        bookingStatus: BookingStatus,
        cancellationDate: LocalDateTime,
    ): List<Booking> =
        bookingJpaRepository.findByBookingStatusAndCancellationDateBefore(bookingStatus, cancellationDate)
            .map(bookingPersistenceMapper::toDomain)

    override fun replaceGuestPassword(userId: Long, encodedPassword: String): Int =
        bookingJpaRepository.replaceGuestPassword(userId, encodedPassword)

    override fun findByUserId(userId: Long?): List<Booking> =
        bookingJpaRepository.findByUserId(userId).map(bookingPersistenceMapper::toDomain)

    override fun existsActiveBookingByScheduleIds(
        scheduleIds: List<Long>,
        excludedStatuses: List<BookingStatus>,
    ): Boolean {
        if (scheduleIds.isEmpty()) {
            return false
        }
        return bookingJpaRepository.existsActiveBookingByScheduleIds(scheduleIds, excludedStatuses)
    }

    override fun deleteInactiveBookingsByScheduleIds(
        scheduleIds: List<Long>,
        inactiveStatuses: List<BookingStatus>,
    ): Int {
        if (scheduleIds.isEmpty()) {
            return 0
        }
        return bookingJpaRepository.deleteInactiveBookingsByScheduleIds(scheduleIds, inactiveStatuses)
    }
}
